cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewManager.java

```diff
@@ -154,32 +157,66 @@ public void onRemoval(RemovalNotification<ApplicationId, Injector> notification)
             if (runner instanceof Service) {
               stopQuietly((Service) runner);
             }
+            removePreviewDir(injector.getInstance(Key.get(ProgramId.class,
+                                                          Names.named(PreviewRunnerModule.PREVIEW_PROGRAM_ID))));
           }
-          ApplicationId application = notification.getKey();
-          if (application == null) {
-            return;
-          }
-          removePreviewDir(application);
         }
       })
       .build();
   }
 
+  @Override
+  protected void startUp() throws Exception {
+    File previewDir = previewDataDir.toFile();
+
+    File[] files = previewDir.listFiles();
+    if (files != null) {
+      // there should be at most 10 directories so the process should not take very long
+      for (File file : files) {
+        ProgramId programId;
+        try {
+          programId = ProgramId.fromString(file.getName());
+        } catch (Exception e) {
+          // if there is an exception converting to a preview id, just continue
+          continue;
+        }
+        Injector injector = createPreviewInjector(programId);
+        PreviewRunner runner = injector.getInstance(PreviewRunner.class);
+        if (runner instanceof Service) {
+          ((Service) runner).startAndWait();
+        }
+        appInjectors.put(programId.getParent(), injector);
```

Comment:
Why not change to cache by program id?

Reply:
This is because the current api is based on the application id, a http call will only pass in the application id, so I still keep app id as the key